File: public/methods/10_FORECAST_EVALUATION/QM004_DIEBOLD_MARIANO_TEST/labs/python/qm004_hands_on.py
"""Reader-facing QM004 hands-on lab."""
from __future__ import annotations
from pathlib import Path
import argparse, math
import pandas as pd
import numpy as np

HERE=Path(__file__).resolve(); ROOT=HERE.parents[2]
DATA=ROOT/'figures/source_data/qm004_forecast_errors.csv'

def p2(z): return math.erfc(abs(z)/math.sqrt(2.0))

def calc(loss: str, lag: int):
    df=pd.read_csv(DATA)
    ea=df.error_a.to_numpy(float); eb=df.error_b.to_numpy(float)
    if loss=='squared': d=ea**2-eb**2
    else: d=np.abs(ea)-np.abs(eb)
    n=len(d); x=d-d.mean(); lrv=float(np.dot(x,x)/n)
    for k in range(1,lag+1): lrv += 2*float(np.dot(x[:-k],x[k:])/n)
    if lrv <= 0: raise ValueError('Estimated long-run variance is non-positive for this choice.')
    stat=float(d.mean()/math.sqrt(lrv/n))
    return {'mean_loss_diff':float(d.mean()),'long_run_variance':lrv,'dm_stat':stat,'two_sided_p':p2(stat)}

def main():
    ap=argparse.ArgumentParser(); ap.add_argument('--loss',choices=['squared','absolute'],default='squared'); ap.add_argument('--lag',type=int,default=3)
    a=ap.parse_args();
    if a.lag < 0: raise SystemExit('--lag must be non-negative')
    r=calc(a.loss,a.lag)
    print(f"Loss: {a.loss}; autocovariance truncation lag: {a.lag}")
    for k,v in r.items(): print(f"{k}: {v:.6f}")
    for k,v in r.items(): print(f"QM004_LAB_METRIC {k}={v:.15g}")

if __name__=='__main__': main()

File: public/methods/10_FORECAST_EVALUATION/QM004_DIEBOLD_MARIANO_TEST/labs/r/qm004_hands_on.R
args <- commandArgs(trailingOnly=TRUE)
loss <- if (length(args)>=1) args[1] else "squared"
lag <- if (length(args)>=2) as.integer(args[2]) else 3
script_arg <- grep("^--file=", commandArgs(FALSE), value=TRUE)
script_path <- if (length(script_arg)) sub("^--file=", "", script_arg[1]) else "labs/r/qm004_hands_on.R"
root <- normalizePath(file.path(dirname(script_path), "../.."), mustWork=TRUE)
d <- read.csv(file.path(root,"figures/source_data/qm004_forecast_errors.csv"))
if (loss == "squared") diff <- d$error_a^2 - d$error_b^2 else if (loss == "absolute") diff <- abs(d$error_a)-abs(d$error_b) else stop("loss must be squared or absolute")
n <- length(diff); x <- diff-mean(diff); lrv <- sum(x*x)/n
if (lag > 0) for (k in 1:lag) lrv <- lrv + 2*sum(x[1:(n-k)]*x[(k+1):n])/n
if (lrv <= 0) stop("Estimated long-run variance is non-positive")
stat <- mean(diff)/sqrt(lrv/n); p <- 2*pnorm(-abs(stat))
cat(sprintf("Loss: %s; autocovariance truncation lag: %d\n",loss,lag))
cat(sprintf("mean_loss_diff: %.6f\nlong_run_variance: %.6f\ndm_stat: %.6f\ntwo_sided_p: %.6f\n",mean(diff),lrv,stat,p))
cat(sprintf("QM004_LAB_METRIC dm_stat=%.12f\n",stat))
cat(sprintf("QM004_LAB_METRIC two_sided_p=%.12f\n",p))
